<launch>
  <arg name="uuv_name" />
  <arg name="joy_id" default="0"/>
  <arg name="use_param_file" default="false"/>
  <arg name="filename" default="."/>

  <!-- Input arguments to map the joystick, default values for XBox 360 controller -->
  <arg name="axis_thruster" default="1"/>
  <arg name="axis_roll" default="0"/>
  <arg name="axis_pitch" default="4"/>
  <arg name="axis_yaw" default="3"/>

  <!--
  Arguments to describe the fin configuration, only used if no parameter
  file is available, for this default mode the thruster is modelled as
    tau = rotorGain * angVel * abs(angVel)
  -->
  <arg name="thruster_rotor_gain" default="0.0009"/>
  <arg name="max_thrust" default="200"/>
  <arg name="thruster_topic" default="thrusters/id_0/input"/>
  <arg name="fin_topic_prefix" default="fins/"/>
  <arg name="fin_topic_suffix" default="/input"/>
  <arg name="thruster_joy_gain" default="1.0"/>
  <!--
  The default fin configuration considered here assumes that your fins are placed
  around the body of the vehicle as follows (on the YZ plane)
  Fin 0: (-y, z)
  Fin 1: (-y -z)
  Fin 2: (y, -z)
  Fin 3: (y, z)
  If you want to pass these lists on the command line, remove the empty spaces
  between the number
  -->
  <arg name="n_fins" default="4"/>
  <arg name="gain_roll" default="1,1,1,1"/>
  <arg name="gain_pitch" default="1,1,-1,-1"/>
  <arg name="gain_yaw" default="-1,1,1,-1"/>

  <!-- If option to use a parameter file is given -->
  <group if="$(var use_param_file)">
    <group>
      <push-ros-namespace namespace="$(var uuv_name)"/>
      <node pkg="uuv_teleop" exec="finned_uuv_teleop.py" name="finned_uuv_teleop"
            output="screen">
        <param from="$(var filename)"/>
      </node>

      <node pkg="joy" exec="joy_node" name="joystick">
        <param name="autorepeat_rate" value="10" />
        <param name="dev" value="/dev/input/js$(var joy_id)"/>
      </node>
    </group>
  </group>

  <group unless="$(var use_param_file)">
    <group>
      <push-ros-namespace namespace="$(var uuv_name)"/>
      <node pkg="uuv_teleop" exec="finned_uuv_teleop.py" name="finned_uuv_teleop" output="screen">
        <param name="axis_thruster" value="$(var axis_thruster)"/>
        <param name="axis_roll" value="$(var axis_roll)"/>
        <param name="axis_pitch" value="$(var axis_pitch)"/>
        <param name="axis_yaw" value="$(var axis_yaw)"/>
        <param name="n_fins" value="$(var n_fins)"/>
        <param name="thruster_joy_gain" value="$(var thruster_joy_gain)"/>
        <param name="thruster_model">
          <param name="name" value="proportional"/>
          <param name="max_thrust" value="$(var max_thrust)"/>
          <param name="params">
            <param name="gain" value="$(var thruster_rotor_gain)"/>
          </param>
        </param>
        
        <!-- Use type="list_of_float" in foxy for lists ?-->
        <param name="gain_roll" value="[$(var gain_roll)]"/>
        <param name="gain_pitch" value="[$(var gain_pitch)]"/>
        <param name="gain_yaw" value="[$(var gain_yaw)]"/>
        <param name="thruster_topic" value="$(var thruster_topic)"/>
        <param name="fin_topic_prefix" value="$(var fin_topic_prefix)"/>
        <param name="fin_topic_suffix" value="$(var fin_topic_suffix)"/>
      </node>

      <node pkg="joy" exec="joy_node" name="joystick">
        <param name="autorepeat_rate" value="10.0" />
        <param name="dev" value="/dev/input/js$(var joy_id)"/>
      </node>
    </group>
  </group>

</launch>
